File: facade/src/main/java/org/tdf/sunflower/consensus/poa/PoaUtils.java
package org.tdf.sunflower.consensus.poa;

import org.tdf.common.util.HashUtil;
import org.tdf.rlpstream.Rlp;
import org.tdf.sunflower.types.Header;

public class PoaUtils {

    public static byte[] getRawHash(Header header) {
        byte[] encoded = Rlp.encode(new Object[] {
            header.getHashPrev(), header.getUnclesHash(), header.getCoinbase(), header.getStateRoot(),
            header.getTransactionsRoot(), header.getReceiptTrieRoot(), header.getLogsBloom(), header.getDifficulty(),
            header.getHeight(), header.getGasLimit(), header.getGasUsed(), header.getCreatedAt(),
            header.getMixHash(), header.getNonce()
        });
        return HashUtil.sha3(encoded);
    }
}

File: facade/src/main/java/org/tdf/sunflower/consensus/poa/PoAUtils.kt
package org.tdf.sunflower.consensus.poa

import org.tdf.common.crypto.ECKey
import org.tdf.common.util.*
import org.tdf.sunflower.types.Header
import org.tdf.sunflower.types.HeaderImpl.Companion.emptyNonce

object PoAUtils {
    fun getRawHash(header: Header): ByteArray {
        return header.impl.copy(extraData = HexBytes.empty(), mixHash = ByteUtil.ZEROS_32, nonce = emptyNonce).rlp()
            .sha3()
    }

    fun sign(key: ECKey, header: Header): Header {
        val hash = getRawHash(header)
        val sig = key.sign(hash)
        val l = java.lang.Byte.toUnsignedLong(sig.v)
        return header.impl.copy(
            extraData = sig.r.bytes32().hex(),
            mixHash = sig.s.bytes32().hex(),
            nonce = l.bytes8().hex()
        )
    }
}